### ISSUE TITLE
Packages.yaml variants: ^something@x.y.z does nothing

### ISSUE THREAD

**Issue by abouteiller** [spack/spack#11001]
I am trying to set a systemwide preference for building `gdb` with `python3` as the python provider.  However, setting the concretization preference in the `etc/spack/packages.yaml file has no effect. 

### Steps to reproduce the issue

I can obviously achieve the stated goal on the command line with `spack install gdb ^python@3.7.2`. That works fine.

However, setting it in the `etc/spack/packages.yaml` file has no effect. 
```yaml
packages:
  gdb:
    variants: ^python@3.7.2
```

Note that setting other concretization preferences (like -xz) in `variants` works.
I also tried to set `providers` but this has no effect either (and if I understand correctly this is normal).

### Error Message

`gdb@8.2.0^python@2.7.16` is produced instead of `gdb@8.2.0^python@3.7.2`

### Information on your system

 1. Scientific Linux 7
 2. ~~any relevant configuration detail (custom `packages.yaml` or `modules.yaml`, etc.)~~ see above

**Comment by mwkrentel** [comment]
I don't know that this syntax is supported.  What is supported is
specifying the python version as an option to python, not to gdb.
In `packages.yaml`,

```
packages:
  ...
  python:
    version: [3.7.2]
```

Note: gdb has a variant `python`, which if True, then adds the
dependency on python.  But the variant is just a boolean, `+python` or
`~python`, it doesn't have a version number.  They're both named
'python', but they're different.

The default for python for gdb is True, but if you want to be explicit
about it, then you would add:

```
  gdb:
    version: [ ... ]
    variants: +python
```

The reason it worked on the command line is because `^python` is the
syntax for a dependency, not a varaint.

So, this is not really a bug.

**Comment by abouteiller** [comment]
I see. However, setting the variant globally on Python has side effects on a number of other packages and is generally undesirable. Is there any correct way of setting the dependency in the package.yaml file?

**Comment by mwkrentel** [comment]
Yes, I can suggest one way that is supported.  Clearly, what you need
are multiple `package.yaml` files, one for each config.  So, the
question is, how do you manage multiple files with the same name.

The solution is to create a separate directory for each config, each
with their own `package.yaml` file.  Use the `spack -C` option to
specify the directory.  (Note: `-C` is an option to spack, not the
install subcommand.)

```
spack -C my-config-dir install package ...
```

Each directory is a full config dir.  You can include `packages.yaml`,
`config.yaml`, `modules.yaml`, anything you want.  Just remember to
always use `-C dir`, or else you'll get inconsistent results.

https://spack.readthedocs.io/en/latest/configuration.html#configuration-scopes

I may be selling spack short.  It's possible there may be other ways
of specifying dependencies that I don't know about.  So, it's worth
reading the docs and fiddling with things to see what works.  You can
always run `spack spec package` to see how spack would build something
without actually having to install the whole thing.

Also, spack environments (a work in progress, I believe) are a way of
bundling several packages into a coherent unit.  But this is something
I know little about.

**Comment by becker33** [comment]
This is supported via:

```
packages:
  gdb:
    require: "%python@3.7.2"
```
and some equivalent forms of specifying the same thing in packages.yaml.
